### ISSUE TITLE
`torch.nested.nested_tensor_from_jagged` does not automatically compute `max_seqlen`

### ISSUE THREAD

**Issue by alanhdu** [pytorch/pytorch#159380]
### 🐛 Describe the bug

If I run the code:

```python
import torch

values = torch.as_tensor([2, 3, 3, 4])
nt1 = torch.nested.as_nested_tensor(
    [values[:1], values[1:]], layout=torch.jagged
)
nt2 = torch.nested.nested_tensor_from_jagged(
    values, lengths=torch.as_tensor([1, 3])
)
nt3 = torch.nested.nested_tensor_from_jagged(
    values, lengths=torch.as_tensor([1, 3]), max_seqlen=3
)

print(nt1.to_padded_tensor(0).shape)
print(nt2.to_padded_tensor(0).shape)
print(nt3.to_padded_tensor(0).shape)
```

`nt1` and `nt3` have shapes `(2, 3)`, but `nt2` has a shape `(2, 4)`, with an unnecessary extra padded element!

I am confused why there is this extra padding element at all, and why manually specifying `max_seqlen` seems to solve the issue.

### Versions

```
Collecting environment information...
PyTorch version: 2.7.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Fedora Linux 41 (Workstation Edition) (x86_64)
GCC version: (GCC) 14.3.1 20250523 (Red Hat 14.3.1-1)
Clang version: 19.1.7 (Fedora 19.1.7-4.fc41)
CMake version: Could not collect
Libc version: glibc-2.40

Python version: 3.12.11 | packaged by conda-forge | (main, Jun  4 2025, 14:45:31) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.15.4-100.fc41.x86_64-x86_64-with-glibc2.40
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  12
On-line CPU(s) list:                     0-11
Vendor ID:                               GenuineIntel
Model name:                              13th Gen Intel(R) Core(TM) i7-1365U
CPU family:                              6
Model:                                   186
Thread(s) per core:                      2
Core(s) per socket:                      10
Socket(s):                               1
Stepping:                                3
CPU(s) scaling MHz:                      39%
CPU max MHz:                             5200.0000
CPU min MHz:                             400.0000
BogoMIPS:                                5376.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               352 KiB (10 instances)
L1i cache:                               576 KiB (10 instances)
L2 cache:                                6.5 MiB (4 instances)
L3 cache:                                12 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-11
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Mitigation; Clear Register File
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] torch==2.7.1+cpu
[conda] Could not collect
```

**Comment by jbschlosser** [comment]
Thanks for the report! I definitely agree the situation results in unexpected / incorrect results. Note that unfortunately we have deprioritized this work, but I will provide some context on the details and perhaps one of you would be willing to propose a workable fix.

> I am lacking context to judge whether this if-clause is necessary. But if inp._max_seqlen_tensor has not been initialized, it will set max_seq_len to the sum of all nested tensor lengths, which is clearly wrong. I would assume this line should simply be max_seq_len = inp._max_seqlen, since inp._max_seqlen will automatically be initialized if it is None.

The crux of the issue comes from the fact that min / max sequence length calculation has the following performance issues:
1. DtoH sync on access (generally bad for performance) with GPU-side offsets / lengths
2. Data-dependency-related graph break within `torch.compile`

So we try to avoid doing this implicitly whenever possible. In fact, I'd go so far as to say that the performance-minded should generally not rely on lazy computation of these values, but rather compute them ahead-of-time and provide them manually. Note that `nested_tensor_from_jagged()` accepts min / max sequence length as optional args.

> But if inp._max_seqlen_tensor has not been initialized, it will set max_seq_len to the sum of all nested tensor lengths, which is clearly wrong

```python
max_seq_len = (
    inp._max_seqlen
    if inp._max_seqlen_tensor is not None
    else inp._values.size(0)
)
```

Agreed that this is clearly wrong, but it's an upper bound that still works with fallback implementations of NJT ops in the form of:
1. convert NJT -> padded
2. perform op on padded
3. convert padded result -> NJT

So we made a tradeoff here to avoid the above performance footguns for ops we've implemented in this way.

It's a pretty messy situation now. I think if we were to start over, perhaps we would entirely remove lazy computation of min / max sequence length and make everything explicit (e.g. provide an option during NJT construction for computing these and error out anywhere they're expected when not available).
